### PR TITLE
Vfio interrupt support

### PR DESCRIPTION
**Pull request by tzwickl**
Add VFIO interrupt support for ixy driver.

### CODE REVIEW

File: src/app/ixy-fwd.c [hunk 1]
```diff
@@ -32,8 +33,8 @@ int main(int argc, char* argv[]) {
 		return 1;
 	}
 
-	struct ixy_device* dev1 = ixy_init(argv[1], 1, 1);
-	struct ixy_device* dev2 = ixy_init(argv[2], 1, 1);
+	struct ixy_device* dev1 = ixy_init(argv[1], 1, 1, 0x028, -1, true);
+	struct ixy_device* dev2 = ixy_init(argv[2], 1, 1, 0, 0, false);
```
Comment (emmericp) [inline src/app/ixy-fwd.c]: setting the ITR should be a separate function because it's specific to intel NICs

File: src/driver/ixgbe.c [hunk 2]
```diff
@@ -51,6 +56,213 @@ struct ixgbe_tx_queue {
...
 };
 
+/**
+ * ixgbe_set_ivar - set the IVAR registers, mapping interrupt causes to vectors
+ * @param dev pointer to device
+ * @param direction 0 for Rx, 1 for Tx
+ * @param queue queue to map the corresponding interrupt to
+ * @param msix_vector the vector to map to the corresponding queue
+ */
+static void set_ivar(struct ixgbe_device *dev, s8 direction, u8 queue, u8 msix_vector) {
```
Comment (emmericp) [inline src/driver/ixgbe.c]: please use C++/Stroustrup style for pointers: `foo* bar`, not `foo *bar` to be consistent with existing code

Comment (emmericp) [inline src/driver/ixgbe.c]: use standard int types from stdint.h

File: src/driver/ixgbe.c [hunk 3]
```diff
@@ -318,18 +544,24 @@ struct ixy_device* ixgbe_init(const char* pci_addr, uint16_t rx_queues, uint16_t
...
+		// initialize interrupts for this device
+		setup_interrupts(dev);
 	} else {
 		debug("mapping BAR0 region via pci file...");
 		dev->addr = pci_map_resource(pci_addr);
 	}
 	dev->rx_queues = calloc(rx_queues, sizeof(struct ixgbe_rx_queue) + sizeof(void*) * MAX_RX_QUEUE_ENTRIES);
 	dev->tx_queues = calloc(tx_queues, sizeof(struct ixgbe_tx_queue) + sizeof(void*) * MAX_TX_QUEUE_ENTRIES);
 	reset_and_init(dev);
+
```
Comment (emmericp) [inline src/driver/ixgbe.c]: remove

File: src/driver/ixgbe.c [hunk 4]
```diff
@@ -381,13 +613,24 @@ void ixgbe_read_stats(struct ixy_device* ixy, struct device_stats* stats) {
...
+	struct interrupt_queues *interrupt = NULL;
+	bool interrupts_enabled = ixy->interrupts.interrupts_enabled;
+
+	if (interrupts_enabled) {
+		interrupt = &ixy->interrupts.queues[queue_id];
+	}
+
+	if (interrupts_enabled && interrupt->interrupt_enabled) {
+		vfio_epoll_wait(interrupt->vfio_epoll_fd, 10, dev->ixy.interrupts.timeout_ms);
+    }
```
Comment (emmericp) [inline src/driver/ixgbe.c]: check tabs vs. whitespace indentation, should be all tabs in ixy

File: src/interrupts.c [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Calculate packets per second based on the received number of packets and the elapsed time in nanoseconds since the
+ * last calculation.
+ * @param received_pkts Number of received packets.
+ * @param elapsed_time_nanos Time elapsed in nanoseconds since the last calculation.
+ * @return Packets per second.
+ */
+static double mpps(uint64_t received_pkts, uint64_t elapsed_time_nanos) {
+	return (double) received_pkts / 1000000.0 / ((double) elapsed_time_nanos / 1000000000.0);
+}
```
Comment (emmericp) [inline src/interrupts.c:17]: can this be done without using floating point?

File: src/interrupts.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	double average = avg->sum / avg->length;
+	if (average > INTERRUPT_THRESHOLD) {
+		interrupt->interrupt_enabled = false;
+	} else if (buf_index == buf_size) {
+		interrupt->interrupt_enabled = false;
+	} else {
+		interrupt->interrupt_enabled = true;
+	}
+	interrupt->last_time_checked = monotonic_time();
+}
```
Comment (emmericp) [inline src/interrupts.c]: please end files with new lines (C99 actually requires this)

File: src/libixy-vfio.c [hunk 7]
```diff
@@ -33,8 +39,188 @@ void vfio_enable_dma(int device_fd) {
...
+	struct vfio_irq_set *irq_set;
+	int *fd_ptr;
+
+	// setup event fd
+	int event_fd = eventfd(0, 0);
+
+	irq_set = (struct vfio_irq_set *) irq_set_buf;
+	irq_set->argsz = sizeof(irq_set_buf);;
+	if (!interrupt_vector)
+		interrupt_vector = 1;
```
Comment (emmericp) [inline src/libixy-vfio.c:107]: please always use braces, even for one-line blocks

File: src/libixy-vfio.c [hunk 8]
```diff
@@ -33,8 +39,188 @@ void vfio_enable_dma(int device_fd) {
...
+ * @param device_fd The VFIO file descriptor.
+ * @return 0 on success.
+ */
+int vfio_disable_msix(int device_fd) {
+	info("Disable MSIX Interrupts");
+	struct vfio_irq_set *irq_set;
+	char irq_set_buf[MSIX_IRQ_SET_BUF_LEN];
+
+	irq_set = (struct vfio_irq_set *) irq_set_buf;
+	irq_set->argsz = sizeof(struct vfio_irq_set);;
```
Comment (emmericp) [inline src/libixy-vfio.c]: double ;

File: src/stats.h [hunk 9]
```diff
@@ -1,6 +1,8 @@
 #ifndef IXY_STATS_H
 #define IXY_STATS_H
 
+#define S_TO_NS(sec) (sec * 1000 * 1000 * 1000)
```
Comment (emmericp) [inline src/stats.h]: only used once, maybe remove?

File: src/driver/ixgbe.c [hunk 10]
```diff
@@ -432,6 +675,26 @@ uint32_t ixgbe_rx_batch(struct ixy_device* ixy, uint16_t queue_id, struct pkt_bu
 		set_reg32(dev->addr, IXGBE_RDT(queue_id), last_rx_index);
 		queue->rx_index = rx_index;
 	}
+
+	if (interrupts_enabled) {
+		bool int_en = interrupt->interrupt_enabled;
+		interrupt->rx_pkts += buf_index;
+
+		uint64_t diff = monotonic_time() - interrupt->last_time_checked;
```
Comment (emmericp) [inline src/driver/ixgbe.c]: i don't like running monotonic_time for every poll; sure, it should be a simple vdso call that should just be rdtsc and it should be fast.

but can we quantify the overhead here? i.e. run a performance evaluation at a lower clock speed and compare interrupts enabled but dynamically disabled due to high load vs. interrupts completely disabled at initilization?

**Review by emmericp — CHANGES_REQUESTED** [review]

File: src/app/ixy-fwd.c (hunk 1 above)
Comment (tzwickl) [reply src/app/ixy-fwd.c]: Ok what I have done now is to remove the parameter and move it into the `ixgbe` driver.

File: src/driver/ixgbe.c (hunk 10 above)
Comment (tzwickl) [reply src/driver/ixgbe.c]: ok I will try to find a better solution for that then using `monotonic_time` :)

File: src/interrupts.c (hunk 5 above)
Comment (tzwickl) [reply src/interrupts.c:17]: Because of performance? Should I also change the average moving then to integers only? I will try to convert it to pps instead of mpps :)

File: src/driver/ixgbe.c (hunk 10 above)
Comment (emmericp) [reply src/driver/ixgbe.c]: not saying monotonic_time is necessarily wrong, a benchmark would be interesting.

for your thesis: how does the kernel ixgbe driver handle this? is it only based on last call there?

would it be possible to only check the time every X calls?

Comment (tzwickl) [reply src/driver/ixgbe.c]: I changed it now to only check the monotonic_time every X calls and now there is no visible performance degradation as if interrupts are disabled :) Before there was a 2-5% increase in CPU utilization compared to now.
For my thesis I will check how the kernel driver is doing this :)

File: src/interrupts.c (hunk 5 above)
Comment (tzwickl) [reply src/interrupts.c:17]: ok I changed the code now to work without floating points :)

**Comment by emmericp** [comment]
minor issue: we need to check whether we are running with vfio support before trying to enable interrupts. probably best to show a warning if interrupts were requested but vfio is not available and just disable interrupts. this would mean we could keep interrupts enabled in the example without breaking non-iommu setups

File: src/driver/ixgbe.c [hunk 11]
```diff
@@ -507,7 +776,8 @@ uint32_t ixgbe_tx_batch(struct ixy_device* ixy, uint16_t queue_id, struct pkt_bu
 		txd->read.buffer_addr = buf->buf_addr_phy + offsetof(struct pkt_buf, data);
 		// always the same flags: one buffer (EOP), advanced data descriptor, CRC offload, data length
 		txd->read.cmd_type_len =
-			IXGBE_ADVTXD_DCMD_EOP | IXGBE_ADVTXD_DCMD_RS | IXGBE_ADVTXD_DCMD_IFCS | IXGBE_ADVTXD_DCMD_DEXT | IXGBE_ADVTXD_DTYP_DATA | buf->size;
+				IXGBE_ADVTXD_DCMD_EOP | IXGBE_ADVTXD_DCMD_RS | IXGBE_ADVTXD_DCMD_IFCS | IXGBE_ADVTXD_DCMD_DEXT | IXGBE_ADVTXD_DTYP_DATA |
+				buf->size;
```
Comment (emmericp) [inline src/driver/ixgbe.c]: please revert this change

File: src/libixy-vfio.c [hunk 12]
```diff
@@ -47,7 +233,8 @@ int vfio_init(char* pci_addr) {
 	}
 	strncat(path, "iommu_group", sizeof(path) - strlen(path) - 1);
 
-	int len = check_err(readlink(path, iommu_group_path, sizeof(iommu_group_path)), "find the iommu_group for the device");
+	int len = check_err(readlink(path, iommu_group_path, sizeof(iommu_group_path)),
+						"find the iommu_group for the device");
```
Comment (emmericp) [inline src/libixy-vfio.c]: please revert this

File: src/libixy-vfio.c [hunk 13]
```diff
@@ -63,10 +250,12 @@ int vfio_init(char* pci_addr) {
 		set_vfio_container(cfd);
 
 		// check if the container's API version is the same as the VFIO API's
-		check_err((ioctl(cfd, VFIO_GET_API_VERSION) == VFIO_API_VERSION) - 1, "get a valid API version from the container");
+		check_err((ioctl(cfd, VFIO_GET_API_VERSION) == VFIO_API_VERSION) - 1,
+				  "get a valid API version from the container");
```
Comment (emmericp) [inline src/libixy-vfio.c]: please revert

File: src/libixy-vfio.c [hunk 14]
```diff
@@ -63,10 +250,12 @@ int vfio_init(char* pci_addr) {
...
 
 		// check if type1 is supported
-		check_err((ioctl(cfd, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU) == 1) - 1, "get Type1 IOMMU support from the IOMMU container");
+		check_err((ioctl(cfd, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU) == 1) - 1,
+				  "get Type1 IOMMU support from the IOMMU container");
```
Comment (emmericp) [inline src/libixy-vfio.c]: please revert

File: src/libixy-vfio.c [hunk 15]
```diff
@@ -76,7 +265,8 @@ int vfio_init(char* pci_addr) {
 	// check if group is viable
 	struct vfio_group_status group_status = {.argsz = sizeof(group_status)};
 	check_err(ioctl(vfio_gfd, VFIO_GROUP_GET_STATUS, &group_status), "get VFIO group status");
-	check_err(((group_status.flags & VFIO_GROUP_FLAGS_VIABLE) > 0) - 1, "get viable VFIO group - are all devices in the group bound to the VFIO driver?");
+	check_err(((group_status.flags & VFIO_GROUP_FLAGS_VIABLE) > 0) - 1,
+			  "get viable VFIO group - are all devices in the group bound to the VFIO driver?");
```
Comment (emmericp) [inline src/libixy-vfio.c]: please revert

File: src/libixy-vfio.c [hunk 16]
```diff
@@ -106,18 +296,20 @@ uint8_t* vfio_map_region(int vfio_fd, int region_index) {
 		// Failed to set iommu type
 		return MAP_FAILED; // MAP_FAILED == ((void *) -1)
 	}
-	return (uint8_t*) check_err(mmap(NULL, region_info.size, PROT_READ | PROT_WRITE, MAP_SHARED, vfio_fd, region_info.offset), "mmap vfio bar0 resource");
+	return (uint8_t*) check_err(mmap(NULL, region_info.size, PROT_READ | PROT_WRITE, MAP_SHARED, vfio_fd, region_info.offset),
+								"mmap vfio bar0 resource");
```
Comment (emmericp) [inline src/libixy-vfio.c]: please revert

File: src/libixy-vfio.c [hunk 17]
```diff
@@ -106,18 +296,20 @@ uint8_t* vfio_map_region(int vfio_fd, int region_index) {
...
-		.vaddr = (uint64_t) vaddr,
-		.iova = iova,
-		.size = size < MIN_DMA_MEMORY ? MIN_DMA_MEMORY : size,
-		.argsz = sizeof(dma_map),
-		.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE};
+			.vaddr = (uint64_t) vaddr,
+			.iova = iova,
+			.size = size < MIN_DMA_MEMORY ? MIN_DMA_MEMORY : size,
+			.argsz = sizeof(dma_map),
+			.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE
```
Comment (emmericp) [inline src/libixy-vfio.c]: please revert

File: src/libixy-vfio.c [hunk 18]
```diff
@@ -126,10 +318,10 @@ uint64_t vfio_map_dma(void* vaddr, uint32_t size) {
...
 uint64_t vfio_unmap_dma(int fd, uint64_t iova, uint32_t size) {
 	struct vfio_iommu_type1_dma_unmap dma_unmap = {
-		.argsz = sizeof(dma_unmap),
-		.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE,
-		.iova = iova,
-		.size = size
+			.argsz = sizeof(dma_unmap),
+			.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE,
+			.iova = iova,
+			.size = size
```
Comment (emmericp) [inline src/libixy-vfio.c]: please revert

**Review by emmericp — CHANGES_REQUESTED** [review]

**Review by emmericp — APPROVED** [review]
